Auto delete AWS Resources

GJL opened the pull request:
See commits

GJL commented on cleanup_aws.py:
```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""A program to delete stale Jepsen AWS resources created by Terraform.
```
It might be considered risky to just delete the resources. An alternative would be to fail the build and signal manual intervention. Since we have a dedicated test account, I think it's fine to delete resources.

GJL replied:
`jepsen-with-unicorns` is good

GJL replied:
maybe `jepsen-with-autodelete`

zentol commented on terraform/jepsen.tf:
```diff
@@ -173,6 +190,10 @@ resource "aws_iam_role" "default" {
+  }
 }
 
 resource "aws_iam_policy_attachment" "hastorage-default" {
```
Why are these 3 resources not cleaned up?

GJL replied:
These resources cannot be tagged. We delete them when deleting the IAM role, e.g., `delete_policies_from_role(role)`.

zentol commented on cleanup_aws.py:
```diff
@@ -0,0 +1,219 @@
+            result.append(role)
+    return result
+
+def filter_stale_iam_roles(roles):
```
We should be able to further de-duplicate this with `filter_stale_s3_buckets` and `filter_stale_resources`; the only difference is how the tags are extracted.

We could add a new variant that accepts the resource list and a tag extractor; for the latter we would then pass `get_iam_role_tags` etc. .

We will still need this function of course so we can reference it in `delete_resources_commands`, but then it is at least a one-liner.

GJL replied:
Fixed

GJL commented on cleanup_aws.py:
```diff
@@ -0,0 +1,219 @@
+        if resources:
+            logger.debug("Checking Jepsen %s for staleness %s", command['resource_type'], resources)
+        else:
+            logger.debug("No Jepsen %s found", command['resource_type'])
```
I can add `continue`

GJL replied:
Fixed

GJL commented on cleanup_aws.py:
```diff
@@ -0,0 +1,219 @@
+    delete_resources_commands = [
+        {
+            'resource_type': 'EC2 instances',
+            'list': list_jepsen_ec2_instances,
```
I would be ok with `'list_resources'`. The function references are supposed to be verbs.

GJL replied:
Fixed

GJL commented on cleanup_aws.py:
```diff
@@ -0,0 +1,219 @@
+    try:
+        tag_set = bucket.Tagging().tag_set
+        return tags_to_dict(tag_set)
+    except ClientError:
```
Unfortunately that is how the AWS API works; if there are no tags, an exception is thrown: https://stackoverflow.com/a/50377918/9203198

GJL replied:
Fixed

GJL commented:
> The travis failure is quite weird: `./travis-run-tests.sh: line 21: assume-role: command not found`.

Fixed. We do not need `assume-role` anymore.

> Let's also add a comment to `jepsen.tf` that any added `aws` resource should also be added to `cleanup_aws.py`, and a comment to `cleanup_aws.py` that this is related to `jepsen.tf`.

Fixed.

> Overall this looks reasonable; what would be a good way to test this?

Provision some machines with `make apply`. 
then 

```
export AWS_PROFILE=testing
```
````
./cleanup_aws.py
```

This should list some resources but not delete anything (default AWS region should match eu-central-1).

Next, you can wait a bit and set `staleness_threshold_hours` to a low value or set `staleness_threshold_hours` to `-1` and re-run the script (optional: re-run it multiple times).

Finally, run
```
make destroy
```

and assert that the last line printed is:

```
Destroy complete! Resources: 0 destroyed.
```

zentol commented on cleanup_aws.py:
```diff
@@ -4,18 +4,22 @@
 prematurely on the build server. If Terraform fails to clean up, the resources
-will not be deleted at all without manual intervention. For this reason we run
-the program in this module to detect and clean up stale resources from
+woult not not be deleted at all without manual intervention. For this reason
```
```suggestion
would not be deleted at all without manual intervention. For this reason
```

GJL commented:
I will add a `requirements.txt` in the next PR.

GJL commented:
> There seems to be an issue with the date formatting:
> 
> ```
> ValueError: time data '2020-05-19T08:32:04Z' does not match format '%Y-%m-%dT%H:%M:%S%z'
> ```

Python 3.5 does not parse the Z as the +0000 offset.

GJL commented:
Merged.
